In [168]:
import pandas as pd
import numpy as np


In [169]:
df_de = pd.read_csv("../data/Delaware/DE_VTD_Hispanic.csv")
print(f"DE shape: {df_de.shape}")
df_de.head(5)

DE shape: (73, 413)


,Label (Grouping),"01-28, Kent County, Delaware","01-29, Kent County, Delaware","01-30, Kent County, Delaware","01-31, Kent County, Delaware","01-32, Kent County, Delaware","01-33, Kent County, Delaware","01-34, Kent County, Delaware","02-28, Kent County, Delaware","02-29, Kent County, Delaware",...,"08-20, Sussex County, Delaware","08-36, Sussex County, Delaware","08-38, Sussex County, Delaware","08-40, Sussex County, Delaware","08-41, Sussex County, Delaware","09-20, Sussex County, Delaware","09-36, Sussex County, Delaware","10-20, Sussex County, Delaware","10-36, Sussex County, Delaware","11-36, Sussex County, Delaware"
0,Total:,"2,203","11,966","5,543",537,"6,420",866,464,"4,366","8,882",...,878,932,"2,875","1,771","1,644",593,365,383,289,571
1,Hispanic or Latino,209,651,300,20,958,68,53,322,647,...,96,481,157,396,331,55,151,6,39,44
2,Not Hispanic or Latino:,"1,994","11,315","5,243",517,"5,462",798,411,"4,044","8,235",...,782,451,"2,718","1,375","1,313",538,214,377,250,527
3,Population of one race:,"1,894","10,702","4,918",503,"5,049",734,383,"3,772","7,744",...,731,433,"2,626","1,325","1,239",509,185,370,232,508
4,White alone,749,"7,465","3,878",417,"2,182",510,195,"2,697","4,801",...,664,370,"2,462","1,203",912,468,159,355,219,461


In [170]:
df_de = df_de.T
# 1. Turn the 'Label (Grouping)' index into a regular column
df_de = df_de.reset_index()

# 2. Set the columns to the first row's values
df_de.columns = df_de.iloc[0]

# 3. Drop that first row (which is now redundant)
df_de = df_de[1:]

# 4. Clean up the index so it starts at 0 again
df_de = df_de.reset_index(drop=True)

# 5. Remove the '0' label that often sits above the index
df_de.columns.name = None

print(df_de.shape)

(412, 74)


In [171]:
df_de.head(5)

,Label (Grouping),Total:,Hispanic or Latino,Not Hispanic or Latino:,Population of one race:,White alone,Black or African American alone,American Indian and Alaska Native alone,Asian alone,Native Hawaiian and Other Pacific Islander alone,...,American Indian and Alaska Native; Asian; Native Hawaiian and Other Pacific Islander; Some Other Race,Population of five races:,White; Black or African American; American Indian and Alaska Native; Asian; Native Hawaiian and Other Pacific Islander,White; Black or African American; American Indian and Alaska Native; Asian; Some Other Race,White; Black or African American; American Indian and Alaska Native; Native Hawaiian and Other Pacific Islander; Some Other Race,White; Black or African American; Asian; Native Hawaiian and Other Pacific Islander; Some Other Race,White; American Indian and Alaska Native; Asian; Native Hawaiian and Other Pacific Islander; Some Other Race,Black or African American; American Indian and Alaska Native; Asian; Native Hawaiian and Other Pacific Islander; Some Other Race,Population of six races:,White; Black or African American; American Indian and Alaska Native; Asian; Native Hawaiian and Other Pacific Islander; Some Other Race
0,"01-28, Kent County, Delaware","2,203",209,"1,994","1,894",749,"1,080",14,16,0,...,0,0,0,0,0,0,0,0,0,0
1,"01-29, Kent County, Delaware","11,966",651,"11,315","10,702","7,465","2,933",46,188,8,...,0,0,0,0,0,0,0,0,0,0
2,"01-30, Kent County, Delaware","5,543",300,"5,243","4,918","3,878",911,15,71,2,...,0,0,0,0,0,0,0,0,0,0
3,"01-31, Kent County, Delaware",537,20,517,503,417,76,2,8,0,...,0,0,0,0,0,0,0,0,0,0
4,"01-32, Kent County, Delaware","6,420",958,"5,462","5,049","2,182","2,570",42,218,6,...,0,0,0,0,0,0,0,0,0,0


In [172]:
df_de.columns = df_de.columns.str.strip()
columns = df_de.columns
print(columns)

Index(['Label (Grouping)', 'Total:', 'Hispanic or Latino',
       'Not Hispanic or Latino:', 'Population of one race:', 'White alone',
       'Black or African American alone',
       'American Indian and Alaska Native alone', 'Asian alone',
       'Native Hawaiian and Other Pacific Islander alone',
       'Some Other Race alone', 'Population of two or more races:',
       'Population of two races:', 'White; Black or African American',
       'White; American Indian and Alaska Native', 'White; Asian',
       'White; Native Hawaiian and Other Pacific Islander',
       'White; Some Other Race',
       'Black or African American; American Indian and Alaska Native',
       'Black or African American; Asian',
       'Black or African American; Native Hawaiian and Other Pacific Islander',
       'Black or African American; Some Other Race',
       'American Indian and Alaska Native; Asian',
       'American Indian and Alaska Native; Native Hawaiian and Other Pacific Islander',
       'Americ

In [173]:
latino_related = []
black_related = []
alone = []
for col in columns:
    if "latino" in col or "hispanic" in col.lower():
        latino_related.append(col)
    if "black" in col.lower():
        black_related.append(col)
    if "alone" in col.lower():
        alone.append(col)

print(f"latino related: {latino_related}")
print(f"black related: {black_related}")
print(f"alone: {alone}")

latino related: ['Hispanic or Latino', 'Not Hispanic or Latino:']
black related: ['Black or African American alone', 'White; Black or African American', 'Black or African American; American Indian and Alaska Native', 'Black or African American; Asian', 'Black or African American; Native Hawaiian and Other Pacific Islander', 'Black or African American; Some Other Race', 'White; Black or African American; American Indian and Alaska Native', 'White; Black or African American; Asian', 'White; Black or African American; Native Hawaiian and Other Pacific Islander', 'White; Black or African American; Some Other Race', 'Black or African American; American Indian and Alaska Native; Asian', 'Black or African American; American Indian and Alaska Native; Native Hawaiian and Other Pacific Islander', 'Black or African American; American Indian and Alaska Native; Some Other Race', 'Black or African American; Asian; Native Hawaiian and Other Pacific Islander', 'Black or African American; Asian; Some O

In [174]:
columns_needed = ['Label (Grouping)', 'Hispanic or Latino'] + alone
columns_needed

['Label (Grouping)',
 'Hispanic or Latino',
 'White alone',
 'Black or African American alone',
 'American Indian and Alaska Native alone',
 'Asian alone',
 'Native Hawaiian and Other Pacific Islander alone',
 'Some Other Race alone']

In [175]:
df_de = df_de[columns_needed]
df_de

,Label (Grouping),Hispanic or Latino,White alone,Black or African American alone,American Indian and Alaska Native alone,Asian alone,Native Hawaiian and Other Pacific Islander alone,Some Other Race alone
0,"01-28, Kent County, Delaware",209,749,"1,080",14,16,0,35
1,"01-29, Kent County, Delaware",651,"7,465","2,933",46,188,8,62
2,"01-30, Kent County, Delaware",300,"3,878",911,15,71,2,41
3,"01-31, Kent County, Delaware",20,417,76,2,8,0,0
4,"01-32, Kent County, Delaware",958,"2,182","2,570",42,218,6,31
...,...,...,...,...,...,...,...,...
407,"09-20, Sussex County, Delaware",55,468,35,2,4,0,0
408,"09-36, Sussex County, Delaware",151,159,20,0,1,0,5
409,"10-20, Sussex County, Delaware",6,355,8,0,3,1,3
410,"10-36, Sussex County, Delaware",39,219,10,0,0,0,3


In [176]:
cols_to_sum = [
    'White alone', 
    'American Indian and Alaska Native alone', 
    'Asian alone', 
    'Native Hawaiian and Other Pacific Islander alone',
    'Some Other Race alone'
]

for col in cols_to_sum:
    df_de[col] = pd.to_numeric(df_de[col].astype(str).str.replace(',', ''), errors='coerce').fillna(0)

df_de['Others'] = df_de[cols_to_sum].sum(axis=1)
df_de

,Label (Grouping),Hispanic or Latino,White alone,Black or African American alone,American Indian and Alaska Native alone,Asian alone,Native Hawaiian and Other Pacific Islander alone,Some Other Race alone,Others
0,"01-28, Kent County, Delaware",209,749,"1,080",14,16,0,35,814
1,"01-29, Kent County, Delaware",651,7465,"2,933",46,188,8,62,7769
2,"01-30, Kent County, Delaware",300,3878,911,15,71,2,41,4007
3,"01-31, Kent County, Delaware",20,417,76,2,8,0,0,427
4,"01-32, Kent County, Delaware",958,2182,"2,570",42,218,6,31,2479
...,...,...,...,...,...,...,...,...,...
407,"09-20, Sussex County, Delaware",55,468,35,2,4,0,0,474
408,"09-36, Sussex County, Delaware",151,159,20,0,1,0,5,165
409,"10-20, Sussex County, Delaware",6,355,8,0,3,1,3,362
410,"10-36, Sussex County, Delaware",39,219,10,0,0,0,3,222


In [177]:
df_de = df_de.drop(cols_to_sum, axis=1)
df_de.head(5)

,Label (Grouping),Hispanic or Latino,Black or African American alone,Others
0,"01-28, Kent County, Delaware",209,"1,080",814
1,"01-29, Kent County, Delaware",651,"2,933",7769
2,"01-30, Kent County, Delaware",300,911,4007
3,"01-31, Kent County, Delaware",20,76,427
4,"01-32, Kent County, Delaware",958,"2,570",2479


In [178]:
print(df_de.shape)

(412, 4)


#### **Georgia**

In [179]:
df_ga = pd.read_csv("../data/Georgia/GA_VTD_Hispanic.csv")
print(f"GA shape: {df_ga.shape}")
df_ga.head(5)

GA shape: (2699, 76)


,GEO_ID,NAME,P2_001N,P2_002N,P2_003N,P2_004N,P2_005N,P2_006N,P2_007N,P2_008N,...,P2_065N,P2_066N,P2_067N,P2_068N,P2_069N,P2_070N,P2_071N,P2_072N,P2_073N,Unnamed: 75
0,Geography,Geographic Area Name,!!Total:,!!Total:!!Hispanic or Latino,!!Total:!!Not Hispanic or Latino:,!!Total:!!Not Hispanic or Latino:!!Population...,!!Total:!!Not Hispanic or Latino:!!Population...,!!Total:!!Not Hispanic or Latino:!!Population...,!!Total:!!Not Hispanic or Latino:!!Population...,!!Total:!!Not Hispanic or Latino:!!Population...,...,!!Total:!!Not Hispanic or Latino:!!Population...,!!Total:!!Not Hispanic or Latino:!!Population...,!!Total:!!Not Hispanic or Latino:!!Population...,!!Total:!!Not Hispanic or Latino:!!Population...,!!Total:!!Not Hispanic or Latino:!!Population...,!!Total:!!Not Hispanic or Latino:!!Population...,!!Total:!!Not Hispanic or Latino:!!Population...,!!Total:!!Not Hispanic or Latino:!!Population...,!!Total:!!Not Hispanic or Latino:!!Population...,NaN
1,7000000US13001000002,"2, Appling County, Georgia",3563,403,3160,3079,1215,1838,7,10,...,0,0,0,0,0,0,0,0,0,NaN
2,7000000US1300100001B,"1B, Appling County, Georgia",1834,76,1758,1695,1575,109,3,3,...,0,0,0,0,0,0,0,0,0,NaN
3,7000000US1300100001C,"1C, Appling County, Georgia",1538,116,1422,1397,1242,150,0,5,...,0,0,0,0,0,0,0,0,0,NaN
4,7000000US1300100003C,"3C, Appling County, Georgia",2515,263,2252,2183,1528,608,3,39,...,0,0,0,0,0,0,0,0,0,NaN


In [180]:
df_ga.columns = df_ga.iloc[0]
df_ga = df_ga[1:]
df_ga = df_ga.reset_index(drop=True)
df_ga = df_ga.rename(columns={'Geography':'GEOID'})
df_ga.head(5)

,GEOID,Geographic Area Name,!!Total:,!!Total:!!Hispanic or Latino,!!Total:!!Not Hispanic or Latino:,!!Total:!!Not Hispanic or Latino:!!Population of one race:,!!Total:!!Not Hispanic or Latino:!!Population of one race:!!White alone,!!Total:!!Not Hispanic or Latino:!!Population of one race:!!Black or African American alone,!!Total:!!Not Hispanic or Latino:!!Population of one race:!!American Indian and Alaska Native alone,!!Total:!!Not Hispanic or Latino:!!Population of one race:!!Asian alone,...,!!Total:!!Not Hispanic or Latino:!!Population of two or more races:!!Population of five races:,!!Total:!!Not Hispanic or Latino:!!Population of two or more races:!!Population of five races:!!White; Black or African American; American Indian and Alaska Native; Asian; Native Hawaiian and Other Pacific Islander,!!Total:!!Not Hispanic or Latino:!!Population of two or more races:!!Population of five races:!!White; Black or African American; American Indian and Alaska Native; Asian; Some Other Race,!!Total:!!Not Hispanic or Latino:!!Population of two or more races:!!Population of five races:!!White; Black or African American; American Indian and Alaska Native; Native Hawaiian and Other Pacific Islander; Some Other Race,!!Total:!!Not Hispanic or Latino:!!Population of two or more races:!!Population of five races:!!White; Black or African American; Asian; Native Hawaiian and Other Pacific Islander; Some Other Race,!!Total:!!Not Hispanic or Latino:!!Population of two or more races:!!Population of five races:!!White; American Indian and Alaska Native; Asian; Native Hawaiian and Other Pacific Islander; Some Other Race,!!Total:!!Not Hispanic or Latino:!!Population of two or more races:!!Population of five races:!!Black or African American; American Indian and Alaska Native; Asian; Native Hawaiian and Other Pacific Islander; Some Other Race,!!Total:!!Not Hispanic or Latino:!!Population of two or more races:!!Population of six races:,!!Total:!!Not Hispanic or Latino:!!Population of two or more races:!!Population of six races:!!White; Black or African American; American Indian and Alaska Native; Asian; Native Hawaiian and Other Pacific Islander; Some Other Race,NaN
0,7000000US13001000002,"2, Appling County, Georgia",3563,403,3160,3079,1215,1838,7,10,...,0,0,0,0,0,0,0,0,0,NaN
1,7000000US1300100001B,"1B, Appling County, Georgia",1834,76,1758,1695,1575,109,3,3,...,0,0,0,0,0,0,0,0,0,NaN
2,7000000US1300100001C,"1C, Appling County, Georgia",1538,116,1422,1397,1242,150,0,5,...,0,0,0,0,0,0,0,0,0,NaN
3,7000000US1300100003C,"3C, Appling County, Georgia",2515,263,2252,2183,1528,608,3,39,...,0,0,0,0,0,0,0,0,0,NaN
4,7000000US1300100004B,"4B, Appling County, Georgia",1321,62,1259,1216,1147,59,2,5,...,0,0,0,0,0,0,0,0,0,NaN


In [181]:
df_ga.columns = df_ga.columns.astype(str).str.strip()
columns = df_ga.columns
alone = []
for col in columns:
    
    if "alone" in col.lower():
        alone.append(col)
        
print(f"all alone: {alone}")

alone.remove('!!Total:!!Not Hispanic or Latino:!!Population of one race:!!Black or African American alone')
print(f"after: {alone}")


all alone: ['!!Total:!!Not Hispanic or Latino:!!Population of one race:!!White alone', '!!Total:!!Not Hispanic or Latino:!!Population of one race:!!Black or African American alone', '!!Total:!!Not Hispanic or Latino:!!Population of one race:!!American Indian and Alaska Native alone', '!!Total:!!Not Hispanic or Latino:!!Population of one race:!!Asian alone', '!!Total:!!Not Hispanic or Latino:!!Population of one race:!!Native Hawaiian and Other Pacific Islander alone', '!!Total:!!Not Hispanic or Latino:!!Population of one race:!!Some Other Race alone']
after: ['!!Total:!!Not Hispanic or Latino:!!Population of one race:!!White alone', '!!Total:!!Not Hispanic or Latino:!!Population of one race:!!American Indian and Alaska Native alone', '!!Total:!!Not Hispanic or Latino:!!Population of one race:!!Asian alone', '!!Total:!!Not Hispanic or Latino:!!Population of one race:!!Native Hawaiian and Other Pacific Islander alone', '!!Total:!!Not Hispanic or Latino:!!Population of one race:!!Some Othe

In [182]:
columns_needed = (['GEOID', '!!Total:!!Hispanic or Latino', 
                   '!!Total:!!Not Hispanic or Latino:!!Population of one race:!!Black or African American alone']
                  + alone)

df_ga = df_ga[columns_needed]
df_ga.head(5)

,GEOID,!!Total:!!Hispanic or Latino,!!Total:!!Not Hispanic or Latino:!!Population of one race:!!Black or African American alone,!!Total:!!Not Hispanic or Latino:!!Population of one race:!!White alone,!!Total:!!Not Hispanic or Latino:!!Population of one race:!!American Indian and Alaska Native alone,!!Total:!!Not Hispanic or Latino:!!Population of one race:!!Asian alone,!!Total:!!Not Hispanic or Latino:!!Population of one race:!!Native Hawaiian and Other Pacific Islander alone,!!Total:!!Not Hispanic or Latino:!!Population of one race:!!Some Other Race alone
0,7000000US13001000002,403,1838,1215,7,10,0,9
1,7000000US1300100001B,76,109,1575,3,3,0,5
2,7000000US1300100001C,116,150,1242,0,5,0,0
3,7000000US1300100003C,263,608,1528,3,39,0,5
4,7000000US1300100004B,62,59,1147,2,5,1,2


In [183]:
for col in alone:
    df_ga[col] = pd.to_numeric(df_ga[col].astype(str).str.replace(',', ''), errors='coerce').fillna(0)

df_ga['Others'] = df_ga[alone].sum(axis=1)
df_ga = df_ga.drop(alone, axis=1)
df_ga.head(5)

,GEOID,!!Total:!!Hispanic or Latino,!!Total:!!Not Hispanic or Latino:!!Population of one race:!!Black or African American alone,Others
0,7000000US13001000002,403,1838,1241
1,7000000US1300100001B,76,109,1586
2,7000000US1300100001C,116,150,1247
3,7000000US1300100003C,263,608,1575
4,7000000US1300100004B,62,59,1157


In [184]:
df_ga.columns

Index(['GEOID', '!!Total:!!Hispanic or Latino',
       '!!Total:!!Not Hispanic or Latino:!!Population of one race:!!Black or African American alone',
       'Others'],
      dtype='object', name=0)

In [185]:
change = {'!!Total:!!Hispanic or Latino':'Hispanic or Latino', 
          '!!Total:!!Not Hispanic or Latino:!!Population of one race:!!Black or African American alone': 'Black or African American alone'}
df_ga = df_ga.rename(columns=change)
df_ga.head(5)

,GEOID,Hispanic or Latino,Black or African American alone,Others
0,7000000US13001000002,403,1838,1241
1,7000000US1300100001B,76,109,1586
2,7000000US1300100001C,116,150,1247
3,7000000US1300100003C,263,608,1575
4,7000000US1300100004B,62,59,1157


In [186]:
df_ga.shape

(2698, 4)

In [187]:
starts_with_total = df_ga['GEOID'].str.startswith('7000000US')
if False in starts_with_total:
    print("there are id doesn't start with 7000000US")

In [188]:
df_ga['GEOID'] = df_ga['GEOID'].str.replace('7000000US','',regex=False)
df_ga.head(5)

,GEOID,Hispanic or Latino,Black or African American alone,Others
0,13001000002,403,1838,1241
1,1300100001B,76,109,1586
2,1300100001C,116,150,1247
3,1300100003C,263,608,1575
4,1300100004B,62,59,1157


In [189]:
df_ga['GEOID'] = df_ga['GEOID'].astype(str)
df_ga['GEOID'] = df_ga['GEOID'].str[:5] + '-' + df_ga['GEOID'].str[5:]
df_ga.head(5)

,GEOID,Hispanic or Latino,Black or African American alone,Others
0,13001-000002,403,1838,1241
1,13001-00001B,76,109,1586
2,13001-00001C,116,150,1247
3,13001-00003C,263,608,1575
4,13001-00004B,62,59,1157


In [190]:
df_ga = df_ga.sort_values(by='GEOID')
df_ga.head(5)

,GEOID,Hispanic or Latino,Black or African American alone,Others
0,13001-000002,403,1838,1241
1,13001-00001B,76,109,1586
2,13001-00001C,116,150,1247
3,13001-00003C,263,608,1575
4,13001-00004B,62,59,1157


In [191]:
df_precincts = pd.read_csv('GA_precincts.csv')
df_precincts = df_precincts.drop('Unnamed: 0', axis=1)
df_precincts = df_precincts.sort_values(by='GEOID')
df_precincts.head(5)


,GEOID,geometry
4,13001-1B,POLYGON ((-82.29335149742474 31.94083813509874...
5,13001-1C,POLYGON ((-82.45096466524214 31.81324323665519...
3,13001-2,"POLYGON ((-82.28049220920148 31.8493689710918,..."
8,13001-3A,POLYGON ((-82.19945697162436 31.81097094290396...
2,13001-3C,POLYGON ((-82.13346188804348 31.63080080533439...


In [192]:
# temp_df = df_precincts['GEOID'].str.split('-', expand=True)
# temp_df[1] = temp_df[1].str.zfill(6)
# df_precincts['GEOID'] = temp_df[0] + '-' + temp_df[1]



In [193]:
df_precincts = df_precincts.sort_values(by='GEOID')
df_precincts

,GEOID,geometry
4,13001-1B,POLYGON ((-82.29335149742474 31.94083813509874...
5,13001-1C,POLYGON ((-82.45096466524214 31.81324323665519...
3,13001-2,"POLYGON ((-82.28049220920148 31.8493689710918,..."
8,13001-3A,POLYGON ((-82.19945697162436 31.81097094290396...
2,13001-3C,POLYGON ((-82.13346188804348 31.63080080533439...
...,...,...
2673,13321-10,POLYGON ((-83.77576242396358 31.33231854782965...
2670,13321-11,POLYGON ((-83.93015208101691 31.44029495792177...
2674,13321-12,POLYGON ((-83.98256671358905 31.46348322315295...
2679,13321-14,POLYGON ((-83.99672930805114 31.53980036559899...


In [194]:
def clean_to_bones(geoid):
    # 1. Convert to string and uppercase
    geoid = str(geoid).upper()
    # 2. Split by the hyphen
    if '-' in geoid:
        county, vtd = geoid.split('-',1)
    else:
        # If no hyphen, assume first 5 are county (standard for GA)
        county, vtd = geoid[:5], geoid[5:]
    
    # 3. Strip all leading zeros from the VTD part
    vtd_clean = vtd.lstrip('0')
    
    return f"{county}-{vtd_clean}"

# Apply this to BOTH dataframes
df_ga['bone_id'] = df_ga['GEOID'].apply(clean_to_bones)
df_precincts['bone_id'] = df_precincts['GEOID'].apply(clean_to_bones)

In [198]:
num_duplicates = df_ga['bone_id'].duplicated().sum()
print(f"There are {num_duplicates} duplicate IDs in df_ga.")

num_duplicates = df_precincts['bone_id'].duplicated().sum()
print(f"There are {num_duplicates} duplicate IDs in df_precincts.")

There are 0 duplicate IDs in df_ga.
There are 0 duplicate IDs in df_precincts.


In [199]:
df_ga.head(5)

,GEOID,Hispanic or Latino,Black or African American alone,Others,bone_id
0,13001-000002,403,1838,1241,13001-2
1,13001-00001B,76,109,1586,13001-1B
2,13001-00001C,116,150,1247,13001-1C
3,13001-00003C,263,608,1575,13001-3C
4,13001-00004B,62,59,1157,13001-4B


In [200]:
df_precincts.head(5)

,GEOID,geometry,bone_id
4,13001-1B,POLYGON ((-82.29335149742474 31.94083813509874...,13001-1B
5,13001-1C,POLYGON ((-82.45096466524214 31.81324323665519...,13001-1C
3,13001-2,"POLYGON ((-82.28049220920148 31.8493689710918,...",13001-2
8,13001-3A,POLYGON ((-82.19945697162436 31.81097094290396...,13001-3A
2,13001-3C,POLYGON ((-82.13346188804348 31.63080080533439...,13001-3C


In [203]:
merged_df = df_ga.merge(df_precincts, on='bone_id', how='outer')
merged_df.head(5)

,GEOID_x,Hispanic or Latino,Black or African American alone,Others,bone_id,GEOID_y,geometry
0,13001-00001B,76,109,1586.0,13001-1B,13001-1B,POLYGON ((-82.29335149742474 31.94083813509874...
1,13001-00001C,116,150,1247.0,13001-1C,13001-1C,POLYGON ((-82.45096466524214 31.81324323665519...
2,13001-000002,403,1838,1241.0,13001-2,13001-2,"POLYGON ((-82.28049220920148 31.8493689710918,..."
3,NaN,NaN,NaN,NaN,13001-3A,13001-3A,POLYGON ((-82.19945697162436 31.81097094290396...
4,13001-0003A1,24,25,1082.0,13001-3A1,NaN,NaN


In [ ]:
missing_pop = merged_df[merged_df['Hispanic or Latino'].isna()]
missing_pop

,GEOID_x,Hispanic or Latino,Black or African American alone,Others,bone_id,GEOID_y,geometry
3,NaN,NaN,NaN,NaN,13001-3A,13001-3A,POLYGON ((-82.19945697162436 31.81097094290396...
15,NaN,NaN,NaN,NaN,13005-DOUGLAS,13005-DOUGLAS,POLYGON ((-82.59743890907157 31.48377083238479...
17,NaN,NaN,NaN,NaN,13007-ANNA,13007-ANNA,POLYGON ((-84.44203429503072 31.28110532633330...
19,NaN,NaN,NaN,NaN,13007-ELMODEL,13007-ELMODEL,POLYGON ((-84.47440407764945 31.27675434233430...
21,NaN,NaN,NaN,NaN,13007-HOGGARDS MILL,13007-HOGGARDS MILL,POLYGON ((-84.55670352227826 31.25670449179093...
...,...,...,...,...,...,...,...
3607,NaN,NaN,NaN,NaN,13313-TRICHUM (TR),13313-TRICHUM (TR),POLYGON ((-85.08013302956822 34.76450447454872...
3608,NaN,NaN,NaN,NaN,13313-TUNNEL HILL (TH),13313-TUNNEL HILL (TH),POLYGON ((-85.06519279571695 34.81396507485579...
3609,NaN,NaN,NaN,NaN,13313-UPPER TENTH (UT),13313-UPPER TENTH (UT),POLYGON ((-84.84155847819389 34.92141501928059...
3612,NaN,NaN,NaN,NaN,13313-VARNELL (VA),13313-VARNELL (VA),POLYGON ((-84.92093851684768 34.89542799463086...


In [207]:
missing_geo = merged_df[merged_df['geometry'].isna()]
missing_geo

,GEOID_x,Hispanic or Latino,Black or African American alone,Others,bone_id,GEOID_y,geometry
4,13001-0003A1,24,25,1082.0,13001-3A1,NaN,NaN
14,13005-0DOUGL,875,1747,8183.0,13005-DOUGL,NaN,NaN
16,13007-00000A,9,43,139.0,13007-A,NaN,NaN
18,13007-00000E,34,229,205.0,13007-E,NaN,NaN
20,13007-00000H,20,199,197.0,13007-H,NaN,NaN
...,...,...,...,...,...,...,...
3611,13313-0000VA,1423,207,6243.0,13313-VA,NaN,NaN
3614,13313-0000WS,408,108,3148.0,13313-WS,NaN,NaN
3643,13321-000016,25,160,264.0,13321-16,NaN,NaN
3647,13321-000005,23,50,655.0,13321-5,NaN,NaN


In [208]:
print(f"total missing pop: {len(missing_pop)}")
print(f"total missing geo: {len(missing_geo)}")

total missing pop: 954
total missing geo: 972


In [209]:
print(f"len census: {df_ga.shape[0]}")
print(f"len geo: {df_precincts.shape[0]}")

len census: 2698
len geo: 2680


In [ ]:
missing_geo['fips'] = missing_geo['bone_id'].str.split('-', n=1).str[0]
missing_geo['fips'] = pd.to_numeric(missing_geo['fips'], errors='coerce')

# 2. Sort the entire DataFrame by FIPS, then by the bone_id
missing_geo = missing_geo.sort_values(by=['fips', 'bone_id'])
missing_geo

/var/folders/_7/0wq2xjx96vngcxdnd_qxp1z40000gn/T/ipykernel_34676/4145679366.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  missing_geo['fips'] = missing_geo['bone_id'].str.split('-', n=1).str[0]
/var/folders/_7/0wq2xjx96vngcxdnd_qxp1z40000gn/T/ipykernel_34676/4145679366.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  missing_geo['fips'] = pd.to_numeric(missing_geo['fips'], errors='coerce')


,GEOID_x,Hispanic or Latino,Black or African American alone,Others,bone_id,GEOID_y,geometry,fips
4,13001-0003A1,24,25,1082.0,13001-3A1,NaN,NaN,13001
14,13005-0DOUGL,875,1747,8183.0,13005-DOUGL,NaN,NaN,13005
16,13007-00000A,9,43,139.0,13007-A,NaN,NaN,13007
18,13007-00000E,34,229,205.0,13007-E,NaN,NaN,13007
20,13007-00000H,20,199,197.0,13007-H,NaN,NaN,13007
...,...,...,...,...,...,...,...,...
3611,13313-0000VA,1423,207,6243.0,13313-VA,NaN,NaN,13313
3614,13313-0000WS,408,108,3148.0,13313-WS,NaN,NaN,13313
3643,13321-000016,25,160,264.0,13321-16,NaN,NaN,13321
3647,13321-000005,23,50,655.0,13321-5,NaN,NaN,13321


In [213]:
missing_pop['fips'] = missing_pop['bone_id'].str.split('-', n=1).str[0]
missing_pop['fips'] = pd.to_numeric(missing_pop['fips'], errors='coerce')

# 2. Sort the entire DataFrame by FIPS, then by the bone_id
missing_pop = missing_pop.sort_values(by=['fips', 'bone_id'])
missing_pop

/var/folders/_7/0wq2xjx96vngcxdnd_qxp1z40000gn/T/ipykernel_34676/2823163449.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  missing_pop['fips'] = missing_pop['bone_id'].str.split('-', n=1).str[0]
/var/folders/_7/0wq2xjx96vngcxdnd_qxp1z40000gn/T/ipykernel_34676/2823163449.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  missing_pop['fips'] = pd.to_numeric(missing_pop['fips'], errors='coerce')


,GEOID_x,Hispanic or Latino,Black or African American alone,Others,bone_id,GEOID_y,geometry,fips
3,NaN,NaN,NaN,NaN,13001-3A,13001-3A,POLYGON ((-82.19945697162436 31.81097094290396...,13001
15,NaN,NaN,NaN,NaN,13005-DOUGLAS,13005-DOUGLAS,POLYGON ((-82.59743890907157 31.48377083238479...,13005
17,NaN,NaN,NaN,NaN,13007-ANNA,13007-ANNA,POLYGON ((-84.44203429503072 31.28110532633330...,13007
19,NaN,NaN,NaN,NaN,13007-ELMODEL,13007-ELMODEL,POLYGON ((-84.47440407764945 31.27675434233430...,13007
21,NaN,NaN,NaN,NaN,13007-HOGGARDS MILL,13007-HOGGARDS MILL,POLYGON ((-84.55670352227826 31.25670449179093...,13007
...,...,...,...,...,...,...,...,...
3607,NaN,NaN,NaN,NaN,13313-TRICHUM (TR),13313-TRICHUM (TR),POLYGON ((-85.08013302956822 34.76450447454872...,13313
3608,NaN,NaN,NaN,NaN,13313-TUNNEL HILL (TH),13313-TUNNEL HILL (TH),POLYGON ((-85.06519279571695 34.81396507485579...,13313
3609,NaN,NaN,NaN,NaN,13313-UPPER TENTH (UT),13313-UPPER TENTH (UT),POLYGON ((-84.84155847819389 34.92141501928059...,13313
3612,NaN,NaN,NaN,NaN,13313-VARNELL (VA),13313-VARNELL (VA),POLYGON ((-84.92093851684768 34.89542799463086...,13313


In [240]:
problem_count = missing_pop['fips'].value_counts().to_frame('missing_count')
problem_count.head(15)

,missing_count
fips,
13067,148
13121,94
13063,69
13139,31
13045,28
13077,24
13223,21
13217,19
13115,19


In [251]:
# 1. Filter both dataframes for Cobb County
cobb_pop = missing_pop[missing_pop['fips'] == 13117][['bone_id']].reset_index(drop=True)
cobb_geo = missing_geo[missing_geo['fips'] == 13117][['bone_id']].reset_index(drop=True)

# cobb_pop.iloc[13]
# 2. Rename columns so we know which is which
cobb_pop.columns = ['Pop_Bone_ID']
cobb_geo.columns = ['Geo_Bone_ID']

# 3. Concatenate them side-by-side
cobb_comparison = pd.concat([cobb_pop, cobb_geo], axis=1)

# Display the result
cobb_comparison

,Pop_Bone_ID,Geo_Bone_ID
0,13117-34,13117-27
1,13117-35,13117-29
2,13117-36,13117-3
3,13117-38,NaN
4,13117-39,NaN
5,13117-40,NaN
6,13117-41,NaN
7,13117-42,NaN
8,13117-43,NaN
9,13117-44,NaN


FIPS: 13067 -> Take first 2 letter of the first word and the last 2 letter or digit
FIPS: 13139, 13045, 13297 -> Use the number in the parenthesis
FIPS: 13077 -> Remove the word last dash and word after the dash
FIPS: 13223, 13217 -> remove letter in pop
FIPS: 13115 -> use the number in the parenthesis and remove leading 0s
FIPS: 13313,13247 -> use the letters in the parenthesis
FIPS: 13153 -> takes the first string after the first dash
FIPS:  